src/download.jl

```diff
@@ -0,0 +1,142 @@
+        startswith(part, '/') && throw(DomainError(part, "begins with \"/\". $explain"))
+    end
+    joinpath(basepart, parts...)
+end
+
+function try_get_filename_from_headers(headers)
+    content_disp = getkv(headers, "Content-Disposition")
+    if content_disp != nothing
+        # extract out of Content-Disposition line
+        # rough version of what is needed in https://github.com/JuliaWeb/HTTP.jl/issues/179
+        filename_part = match(r"filename\s*=\s*(.*)", content_disp)
+        if filename_part != nothing
+            filename = filename_part[1]
+            quoted_filename = match(r"\"(.*)\"", filename)
+            if quoted_filename != nothing
+                # It was in quotes, so it will be double escaped
+                filename = unescape_string(quoted_filename[1])
+            end
+            return filename
+        end
+    end
+    return nothing
+end
+
+function try_get_filename_from_remote_path(target)
+    target == "" && return nothing
+    filename = basename(target)
+    if filename == ""
+        try_get_filename_from_remote_path(dirname(target))
+    else
+        filename
+    end
+end
+
+
+determine_file(::Nothing, resp) = determine_file(tempdir(), resp)
+# ^ We want to the filename if possible because extension is useful for FileIO.jl
+
+function determine_file(path, resp)
+    # get the name
+    name = if isdir(path)
+        # got to to workout what file to put there
+        filename = something(
+                        try_get_filename_from_headers(resp.headers),
+                        try_get_filename_from_remote_path(resp.request.target),
+                        basename(tempname()) # fallback, basically a random string
+                    )
+        safer_joinpath(path, filename)
+    else
+        # It is a file, we are done.
+        path
+    end
+
+    # get the extension, if we are going to save it in encoded form.
+    if header(resp, "Content-Encoding") == "gzip"
+        name *= ".gz"
+    end
+    name
+end
+
+"""
+    download(url, [local_path], [headers]; update_period=1, kw...)
+
+Similar to `Base.download` this downloads a file, returning the filename.
+If the `local_path`:
+ - is not provided, then it is saved in a temporary directory
+ - if part to a directory is provided then it is saved into that directory
+ - otherwise the local path is uses as the filename to save to.
+
+When saving into a directory, the filename is determined (where possible),
+from the rules of the HTTP.
+
+ - `update_period` controls how often (in seconds) to report the progress.
+    - set to `Inf` to disable reporting
+ - `headers` specifies headers to be used for the HTTP GET request
+ - any additional keyword args (`kw...`) are passed on to the HTTP request.
+"""
+function download(url::AbstractString, local_path=nothing, headers=Header[]; update_period=1, kw...)
+    format_progress(x) = "$(round(100x, digits=2))%"
```

Comment:
Possibly this should not be formatted.
idk what Juno is doing with `progress`.
OhMyLog doesn't mind if it is given a string or a float.